Feature/define hope 3 dependency injection

timothycrosley opened the pull request:
Initial HOPE defining Hugs proposed dependency injection system

timothycrosley committed:
Add beginning stage of dependency injection HOPE

timothycrosley committed:
Expand dependency injection HOPE to further explore what an improved dependency injection system could look like

timothycrosley committed:
Grammer improvements

timothycrosley committed:
Update updated date

timothycrosley committed:
Further specification of depency injection system

timothycrosley committed:
Improve formatting of list in HOPE 3

Peter200lx commented on process/HOPE-3--Dependency-Injection.md:
```diff
@@ -0,0 +1,118 @@
+### Where registered dependency providers are stored
+
+Dependencies providers will be stored within the Hug API module level singleton, within a `dependency_providers` dictionary.
+If a second `provides` function is defined it will simply take the place of the first - similar to defining a second function with the same name.
```
The behavior of overwriting another provider should be a parameter that defaults to `False` I believe. That way the default will be a warning about overwriting, but still allow tests and the like to knowingly opt in to overwriting.

Peter200lx commented on process/HOPE-3--Dependency-Injection.md:
````diff
@@ -0,0 +1,118 @@
+    pass
+```
+
+Arguments can be passed to the `requires` function, these will be passed directly to the `provides` function.
````
Would it be better to have the `args` and `kwargs` parameters that accept a `list` and a `dict` which will be passed into the `provides` decorated function? This way there is less magic happening, and you don't have to worry about parameter name conflicts.

timothycrosley replied:
After some thought, I think it makes sense to pass them along, if and only if, the method in question takes no extra optional arguments. Or rather, if the function signature matches that of `functools.partial` it matches an accepted pattern that has almost no chance of name conflicts, while simplifying usage. If it ever goes beyond this, I think you are 100% correct that more explicit usage should be enforced.

Peter200lx replied:
I think this is a reasonable pattern as long as you can be sure that you will never want to add configuration parameters to the `requires` function. I think it might be worth considering if there's any potential tweaks that might be desired when calling the `requires` function. One example off the top of my head is the idea of potentially renaming a parameter to match a different context:

```python
@hug.provides("mysql")
def mysql_config(host: str, port: int = 3306):
    ...


@hug.cli()
def my_server(
    mysql=requires("mysql", rename={"host": "mysql_host", "port": "mysql_port"}),
    api_host: str,
    api_port: int = 8080,
):
    ...
```
I'm not proposing adding the rename functionality in this HOPE, just highlighting that we'll be preventing such future tweaks if we go with the `functools.partial` pattern. If we're good with making that call, or will have an `advanced_requires` in the future that can be extended I withdraw my concern here.

timothycrosley replied:
I agree 100% with your assertions here! I think I'm okay limiting the requires endpoint to the very limited subset allowed by the `funtools.partials` signature, and then if other uses are deemed common enough creating other functions for them

Peter200lx commented on process/HOPE-3--Dependency-Injection.md:
```diff
@@ -0,0 +1,118 @@
+
+Dependencies providers will be stored within the Hug API module level singleton, within a `dependency_providers` dictionary.
+If a second `provides` function is defined it will simply take the place of the first - similar to defining a second function with the same name.
+If a dependency provider is only meant for a particular set of interfaces, it will be stored on the sub interface API object within the singleton.
```
To confirm then, the behavior of HUG will be to first check the per-interface dict of providers, and only fall through to the global dict if the interface specific one does not have the requested provider?

timothycrosley replied:
Yes that's correct! I've clarified that information in the document as well :)

Peter200lx commented on process/HOPE-3--Dependency-Injection.md:
````diff
@@ -0,0 +1,118 @@
+For instance, in `py.test`'s conftest.py, you could store a series of dependency providers targeting the API you intend to test:
+
+```
+   @hug.provides("mysql", api=production_api_im_testing)
````
If we do have an `override=False` parameter then we'll need to pass `True` to it here.

Peter200lx commented on process/HOPE-3--Dependency-Injection.md:
```diff
@@ -0,0 +1,118 @@
+
+Let's say you reuse a single set of parameters for every endpoint within an API module. Currently, in Hug, the simplest thing to do is redefine these parameters in every function.
+This is inconsistent with the Do It Right ONCE (DRY) principle. In this HOPE we are proposing to allow automatic nesting of any hug decorated endpoint in the same manner as full dependencies,
+albeit without the ability to easily swap them out for testing. Here's the proposed API for this feature:
```
I'll need to think about this more, I'm not sure I agree with hiding parameters from the endpoint. I might be misunderstanding this a little bit.

Peter200lx replied:
Also I think nesting providers might make sense, but not as much nesting endpoints. This seems orthogonal to the rest of this HOPE and I'd encourage nesting endpoints be raised as a separate HOPE if wanted.

Peter200lx reviewed (commented):
Overall I love how this reduces the magic with explicit calls to `requires`.

Here's a few requests to improve the clarity of this document:
* Provide an example in this document of what current functionality of hug this is replacing
* Provide function declarations for the new proposed `provides` / `requires` methods
* Add `python to all the triple tick sections that include code

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md

Add missing python syntax highlighting identifiers

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md

Update HOPE to mention override behaviours

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md

Clarify Hug's behaviour when multiple levels of providers are set.

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md

More explicitly call out new signature and replaced behaviour

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md

Fix typo in old directive documentation

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md

Explain Rejection

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md

Add example of directly requiring functions

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md

Fix name of parameter to match surrounding function within direct requires example

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md

Spelling and grimmer fixes

Peter200lx commented on process/HOPE-3--Dependency-Injection.md:
````diff
@@ -0,0 +1,169 @@
+```python
+@provides(name: str = None, api: hug.API = None, override: bool = False)
+
+def requires(name_or_function: Union[str, Callable], *args, **kwargs) -> Any
````
super minor: As you're demonstrating `provides` in the usage form of `@provides`, I'd recommend dropping the `def ` from the `requires` line. Just the `requires(parameters: types) -> returns` seems good.

timothycrosley replied:
Agreed! Updating :)

Peter200lx commented on process/HOPE-3--Dependency-Injection.md:
````diff
@@ -0,0 +1,169 @@
+These two calls are meant to replace the old `Directive` system:
+
+```python
+@directive(apply_globally: bool = False, api: hug.API = None)
````
This should probably be followed by `def timer(...` so that the "was used" bit below has the `timer` name to connect conceptually. If you want to make it distinguished it might work to have it in a comment:
```python
@directive(apply_globally: bool = False, api: hug.API = None)
#def timer(...):
#   pass
```

timothycrosley replied:
100% agree, I've updated to include the name reference in the directive definition

Peter200lx commented on process/HOPE-3--Dependency-Injection.md:
````diff
@@ -0,0 +1,169 @@
+```python
+    @hug.provides()
+    def mysql_connection(request, config):
+        ...
````
I find myself slightly confused by this example. Is `request` a magic keyword talking about a request object, or is it just the name of a `?request=` or `--request` parameter from the caller? If the latter I'd recommend instead providing an example with `host` and `port` or `user` and `db` because those are more explicitly mysql configuration items.

timothycrosley replied:
Agreed :), I've updated to be host and port. request was a magic variable, however all magic variables should be replaced in Hug 3.0.0 with this dependency system. So if a user needed the request:

```
@hug.provides()
def mysql_connection(request=requires("request")):
     ....
```
would be used instead.

Peter200lx replied:
Makes sense, but we should have a list of reserved `provides` keywords per interface. Or we can just rely on the `override=False` system to warn the user.

Peter200lx commented on process/HOPE-3--Dependency-Injection.md:
````diff
@@ -0,0 +1,169 @@
+The API of the new system would consist of a `provides` decorator for defining available dependencies, and a `requires` function for using them:
+
+```python
+@provides(name: str = None, api: hug.API = None, override: bool = False)
````
As we're in Python 3 I'd recommend forcing the `api` and `override` parameters to be keyword arguments. This can be done with the addition of a `*` parameter:
```python
@provides(name: str = None, *, api: hug.API = None, override: bool = False)
```

timothycrosley replied:
Great call! Updating :)

Peter200lx commented on process/HOPE-3--Dependency-Injection.md:
````diff
@@ -0,0 +1,169 @@
+@hug.http()
+def hello_world(config=requires(shared_configuration)):
+    pass
+```
````
This will require some awesome documentation of API parameters to keep confusion to a minimum. Currently when looking at code I can know function parameter names are interface arguments if there is a `@hug.provides` or `@hug.{cli,http,local}` in front of it. Now any functions parameter naming can magically become runtime critical with no warning. (Looking at `def shared_configuration` as a programmer I have no warning that renaming `config_file_location` to `feature_x_config_file` will break my API interface behavior.)

timothycrosley replied:
This is a good point. On the plus side, in your scenario,  you would get an error immediately by Python itself letting you know that you were referencing a non existent variable in the below `requires` function. It's also a fairly common pattern in python itself to use undecorated functions as callbacks in this way (for example: sort functions). Might think on this a bit longer.

Peter200lx replied:
I don't think you would get the warning immediately. If I call the code from cron once a day with
```
$ http get http://localhost:8000/hello_world?config_file_location=my_config.json
```
and rename `def shared_configuration`'s parameter `config_file_location` to `feature_x_config_file` I won't see the breakage until the next time cron runs.

timothycrosley replied:
Because these requirements are defined at instantiation time, you would see an error when starting up the server, or even just importing the module, rather than at call time: https://pyfiddle.io/fiddle/ce3c1ff2-808d-4800-ac6b-ca5e13e5fb0c/?i=true

Peter200lx replied:
I'm not talking about changing the `requires(shared_configuration)` call, but changing the name of a parameter inside of
```python
def shared_configuration(config_file_location="config_file.json"):
```
to
```python
def shared_configuration(feature_x_config_file="config_file.json"):
```

timothycrosley replied:
Oh sorry for my misunderstanding! Your concern makes perfect sense

Peter200lx replied:
Intelligent editors can trace the changing of a parameter name to find and edit all callers that explicitly set that parameter by keyword. But this change flows out to external interfaces outside of the python code and isn't easy to trace without a thorough understanding of hug and the program at hand.

Peter200lx commented on process/HOPE-3--Dependency-Injection.md:
```diff
@@ -0,0 +1,169 @@
+### Where registered dependency providers are stored
+
+Dependencies providers will be stored within the Hug API module level singleton, within a `dependency_providers` dictionary.
+If a second `provides` function is defined by default it will raise an `ExistingDependency` exception. However, if that second definition defines `override=True`, it will simply take the place of the first - similar to defining a second function with the same name. If a dependency provider is only meant for a particular set of interfaces, it will be stored on the sub interface API object within t ...
```
Should there be any type of warning when defining an interface local `provides` that will shadow a global `provides` of the same name? Alternatively should there be a restriction that you can either have a global `provides` or interface local `provides` for a given name?

timothycrosley replied:
These are great questions! I can definitely see a case where it might make sense to have a default depends and an override for just one interface, but I can also see that being done just as easily on accident. At first I was thinking we could overload `override=True` for any case where there is a global interface to allow creating an interface specific provider, but upon some more thought, I'm thinking that might also lead to unexpected errors as you may want to override the global provider for this interface but not a previous provider defined for it. What if we added an `override_global` parameter only to the per interface provider functions?

Peter200lx commented on process/HOPE-3--Dependency-Injection.md:
```diff
@@ -0,0 +1,169 @@
+### Overriding a dependency
+
+Overriding a dependency in this system is straight forward: You update the dictionary to point to your new dependency providing function. You do this by creating a new provides against the existing provides API singleton, with the `override` parameter set to `True`.
+For instance, in `py.test`'s conftest.py, you could store a series of dependency providers targeting the API you intend to test:
```
minor note: did you mean `test.py`?

Peter200lx replied:
ah, that makes sense. I had heard it called `pytest`, hadn't seen `py.test`.

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md

Add link to conftest.py documentation to make meaning more obvious

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md

Improving signature definition per @Peter200lx's recommendations

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md

Add timer reference to documentation

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md

Improve example mysql provides, removing magic variable references, which should be removed in Hug 3.0.0

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md

Grammer and spelling fixes.

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md

timothycrosley committed:
Update HOPE-3--Dependency-Injection.md
